src/components/EarnDashboard/components/ProviderDetails/OpportunityRow.tsx

```diff
@@ -0,0 +1,208 @@
+} from 'state/slices/opportunitiesSlice/utils'
+import {
+  selectAssetById,
+  selectAssets,
+  selectMarketDataSortedByMarketCap,
+} from 'state/slices/selectors'
+import { useAppSelector } from 'state/store'
+
+import { NestedAsset } from './NestedAsset'
+import { OpportunityRowGrid } from './OpportunityTableHeader'
+
+type OpportunityRowProps<T extends StakingEarnOpportunityType | LpEarnOpportunityType> = {
+  onClick: (opportunity: T, action: DefiAction) => void
+  opportunity: T
+}
+
+export const OpportunityRow: React.FC<
+  OpportunityRowProps<StakingEarnOpportunityType | LpEarnOpportunityType>
+> = ({ onClick, opportunity }) => {
+  const {
+    assetId,
+    underlyingAssetId,
+    fiatAmount,
+    isClaimableRewards,
+    underlyingAssetRatiosBaseUnit,
+    cryptoAmountBaseUnit,
+    underlyingAssetIds,
+    type,
+    apy,
+    icons,
+  } = opportunity
+  const translate = useTranslate()
+  const borderColor = useColorModeValue('blackAlpha.50', 'whiteAlpha.50')
+  const asset = useAppSelector(state => selectAssetById(state, underlyingAssetId))
+  const assets = useAppSelector(selectAssets)
+  const marketData = useAppSelector(selectMarketDataSortedByMarketCap)
+
+  const underlyingAssetLabel = useMemo(() => {
+    if ((opportunity as StakingEarnOpportunityType)?.rewardsCryptoBaseUnit) {
+      return 'common.reward'
+    } else {
+      return 'defi.underlyingAsset'
+    }
+  }, [opportunity])
+
+  const underlyingAssetBalances = useMemo(() => {
+    if ((opportunity as StakingEarnOpportunityType)?.rewardsCryptoBaseUnit) {
+      const earnOpportunity = opportunity as StakingEarnOpportunityType
+      return getRewardBalances({
+        rewardAssetIds: earnOpportunity.rewardAssetIds,
+        rewardsCryptoBaseUnit: earnOpportunity.rewardsCryptoBaseUnit,
+        assets,
+        marketData,
+      })
+    } else {
+      return getUnderlyingAssetIdsBalances({
+        assetId,
+        underlyingAssetIds,
+        underlyingAssetRatiosBaseUnit,
+        cryptoAmountBaseUnit,
+        assets,
+        marketData,
+      })
+    }
+  }, [
+    assetId,
+    assets,
+    cryptoAmountBaseUnit,
+    marketData,
+    opportunity,
+    underlyingAssetIds,
+    underlyingAssetRatiosBaseUnit,
+  ])
+
+  const nestedAssetIds = useMemo(() => {
+    if ((opportunity as StakingEarnOpportunityType)?.rewardsCryptoBaseUnit) {
+      const earnOpportunity = opportunity as StakingEarnOpportunityType
+      return earnOpportunity.rewardAssetIds
+    } else {
+      const lpOpportunity = opportunity as LpEarnOpportunityType
+      return lpOpportunity.underlyingAssetIds
+    }
+  }, [opportunity])
+
+  const handleClick = useCallback(
+    (action: DefiAction) => {
+      onClick(opportunity, action)
+    },
+    [onClick, opportunity],
+  )
+  const subText = [<Amount.Percent value={bnOrZero(apy).toString()} suffix='APY' autoColor />]
+  if (bnOrZero(cryptoAmountBaseUnit).gt(0))
+    subText.push(<RawText textTransform='capitalize'>{type}</RawText>)
```

Comment:
This will push to `subText` every re-render, let's memoize this logic into a JSX expression